### PR TITLE
Add brute force vs bvh benchmark to run multiple dimensions

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: examples/brute_force/brute_force.cpp [hunk 1]
```diff
@@ -25,74 +27,54 @@ struct Dummy
...
+
+  using Box = ArborX::ExperimentalHyperGeometry::Box<DIM>;
 
   for (int i = 0; i < nrepeats; i++)
   {
     unsigned int out_count;
     {
       Kokkos::Timer timer;
-      ArborX::BoundingVolumeHierarchy<MemorySpace> bvh{space, primitives};
+      ArborX::BasicBoundingVolumeHierarchy<MemorySpace, Box> bvh{space,
```
Comment (aprokop) [inline examples/brute_force/brute_force.cpp:76]: Don't like this, but `BVH` and `BoundingVolumeHierarchy` don't have the `BoundingVolume` template parameter. That would have to be cleaned, probably at the same time as APIv2.

**Review by dalg24 — COMMENTED** [review]
Why do you want to change the example in the first place?

**Comment by aprokop** [comment]
> Why do you want to change the example in the first place?

So that we are able to run and profile BVH and BruteForce for higher dimensions. It is also interesting to see when BruteForce is faster than BVH depending on the dimension.

**Comment by aprokop** [comment]
Per discussion with @dalg24, we decided to leave example as it is as a) to not make it overly complex for users, and b) do not expose ArborX internals (like `BasicBoundingVolumeHierarchy`. Instead, a new benchmark was introduced.

**Review by masterleinad — COMMENTED** [review]
Would you mind adding a comment saying that this is basically the same as `examples/brute_force/brute_force.cpp` but for higher dimensions?

**Comment by masterleinad** [comment]
> Would you mind adding a comment saying that this is basically the same as `examples/brute_force/brute_force.cpp` but for higher dimensions?

Ah, I see that you want to modify that file anyway.

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  static KOKKOS_FUNCTION auto get(Dummy<DIM> const &, size_type i)
+  {
+    ArborX::ExperimentalHyperGeometry::Point<DIM> center;
+    for (int d = 0; d < DIM; ++d)
+      center[d] = (float)i;
+    return attach(intersects(ArborX::ExperimentalHyperGeometry::Sphere<DIM>{
+                      center, (float)i}),
+                  i);
+  }
+};
```
Comment (masterleinad) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:61]: Can you please add a comment here describing what's happening here? I know that it's pre-existing but I think it's difficult to figure it out immediately. Just something like "The primitives and predicates are constructed on-the-fly. The ith primitive is a Point with all coordinates i. The ith predicate describes an intersection with a Sphere with center at (i,i,...) and radius i"

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+#include <boost/program_options.hpp>
+
+template <int DIM>
+struct Dummy
+{
+  int count;
+};
+
+using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+using MemorySpace = ExecutionSpace::memory_space;
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:29]: Move them above the definition of `Dummy`

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+template <int DIM>
+struct ArborX::AccessTraits<Dummy<DIM>, ArborX::PrimitivesTag>
+{
+  using memory_space = MemorySpace;
+  using size_type = typename MemorySpace::size_type;
+  static KOKKOS_FUNCTION size_type size(Dummy<DIM> const &d) { return d.count; }
+  static KOKKOS_FUNCTION auto get(Dummy<DIM> const &, size_type i)
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:37]: Pass `Dummy` by value

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  {
+    ArborX::ExperimentalHyperGeometry::Point<DIM> point;
+    for (int d = 0; d < DIM; ++d)
+      point[d] = (float)i;
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:41]: Cast is unnecessary

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    break;
+  case 4:
+    run<4>(nprimitives, nqueries, nrepeats);
+    break;
+  case 5:
+    run<5>(nprimitives, nqueries, nrepeats);
+    break;
+  case 6:
+    run<6>(nprimitives, nqueries, nrepeats);
+    break;
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:162]: How long does it take to compile?

**Review by dalg24 — COMMENTED** [review]
Did you plan on using google benchmark?

**Comment by aprokop** [comment]
> Did you plan on using google benchmark?

I have not thought about it. But it's an interesting idea if we want to run multiple configurations in it.

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp (hunk 5 above)
Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:41]: It does not hurt, and it makes the intent clear. I did not introduce it, but I don't dislike it.

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp (hunk 4 above)
Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:37]: > We overuse the pass-by-const-reference

I find it hard to agree with the overuse. It does not seem like there is guarantee that passing by value if *always* faster than passing by reference. So, it seems like each case has to be considered individually.

Comment (dalg24) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:37]: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:37]: While interesting, it is still not a clear rule of thumb. In #650 we saw that returning a struct with 3 floats was faster by value. What about 5? 7? 10? At what point do we know? This is a general question, I see your point about this specific patch.

Comment (dalg24) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:37]: Just take it by value

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp (hunk 6 above)
Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:162]: On my Mac, compiling the driver with 6 instantiations takes 20s. 1 instantiation takes 7s.

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+using MemorySpace = ExecutionSpace::memory_space;
+
+template <int DIM>
+struct Dummy
+{
+  int count;
+};
+
+// Primitives are a set of points located at (i, i, i),
+// with i = 1, ..., n
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:32]:
```suggestion
// with i = 0, ..., n-1
```

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  int nprimitives;
+  int nqueries;
+  int nrepeats;
+  namespace bpo = boost::program_options;
+  bpo::options_description desc("Allowed options");
+  // clang-format off
+  desc.add_options()
+      ( "help", "help message" )
+      ( "dimension", bpo::value<int>(&dim)->default_value(3), "dimension" )
+      ( "iterations", bpo::value<int>(&nrepeats)->default_value(1), "number of iterations" )
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:132]:
```suggestion
      ( "repetitions", bpo::value<int>(&nrepeats)->default_value(1), "number of repetitions" )
```

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp (hunk 6 above)
Comment (dalg24) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:166]: I am curious whether and by how much it speeds up compilation if we explicitly instantiate in multiple source files

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp (hunk 2 above)
Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp]: Done.

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp (hunk 6 above)
Comment (dalg24) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:166]: quick experiment allowed me to compile 3x faster (see aprokop/ArborX#7)

Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:166]: I tried it out. Got 12s. So about 2x slower than a single instantiation, and about 40% faster than multiple in a single file.

Comment (dalg24) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:166]: Please elaborate, I am not sure I understand what you did nor what the conclusion you drew

Comment (aprokop) [reply benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:166]: I measured the improvements of your suggestion (ETI) on my machine and compared it to the the timings I posted above. I have not drawn any conclusion. I can live with either, as we tradeoff some speedup for the faster compilation vs having a more complicated build and more files.

**Pull request merged**